Email editor - dependencies check [MAILPOET-6367]

costasovo opened the pull request:
## Description

This PR adds a mechanism for checking if a site has the dependencies needed to run the email editor. At this point, we are checking for a minimal WP (6.7).
When the editor can't run, we hide the button for creating an email in the new editor, and in case a user attempts to edit an email previously created in the new editor, they get a notice.

<img width="1017" alt="Screenshot 2025-01-09 at 18 27 14" src="https://github.com/user-attachments/assets/55f7668c-ed29-4423-8bec-7ef6391c7614" />



## Code review notes

Please review commit by commit.

## QA notes

You can use [WP Downgrade](https://wordpress.org/plugins/wp-downgrade/) plugin to install older WP.

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-6367]

## After-merge notes

_N/A_

## Tasks

- [ ] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [ ] I added sufficient test coverage
- [ ] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-6367]: https://mailpoet.atlassian.net/browse/MAILPOET-6367?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Preview

[Preview in WordPress Playground](https://account.mailpoet.com/playground/new/branch:email-editor/feature-check)

_The latest successful build from `email-editor/feature-check` will be used. If none is available, the link won't work._

costasovo committed:
Add service for checking editor dependencies

[MAILPOET-6367]

costasovo committed:
Hide create email in new editor button when deps are not met

[MAILPOET-6367]

costasovo committed:
Redirect to emails and show notice when attempting to open a new email

[MAILPOET-6367]

triple0t commented on packages/php/email-editor/src/Engine/class-dependency-check.php:
```diff
@@ -0,0 +1,47 @@
+	 */
+	private function is_gutenberg_version_compatible(): bool {
+		return version_compare( get_bloginfo( 'version' ), self::MIN_GUTENBERG_VERSION, '>=' );
+	}
```
I believe this should be checking the Gutenberg plugin version. Can you update this section?

costasovo replied:
Addressed in https://github.com/mailpoet/mailpoet/pull/6018/commits/87520fe41d0346191b9035374ce475e4454fd9d9

triple0t reviewed (changes requested):
Thanks for working on this.

I dropped a comment. It appears we aren't checking the Gutenberg plugin version.

Also, when I tried editing an existing newsletter (using WP 6.6.2), I got this error `Invalid post type.`
![Screenshot 2025-01-06 at 3 08 58 PM](https://github.com/user-attachments/assets/675ab26b-9a9a-4a84-9ca8-70f5ebac6250)


Clicking `Preview` leads to a critical error.
![Screenshot 2025-01-06 at 3 14 53 PM](https://github.com/user-attachments/assets/80ef7bd1-b461-4f4e-8dfc-c70cd0635aa2)

Logs
```
[06-Jan-2025 14:12:36 UTC] PHP Fatal error:  Uncaught TypeError: MailPoet\EmailEditor\Engine\Renderer\ContentRenderer\Content_Renderer::render(): Argument #2 ($template) must be of type WP_Block_Template, null given, called in /srv/htdocs/wp-content/plugins/mailpoet/vendor/mailpoet/email-editor/src/Engine/Renderer/class-renderer.php on line 32 and defined in /srv/htdocs/wp-content/plugins/mailpoet/vendor/mailpoet/email-editor/src/Engine/Renderer/ContentRenderer/class-content-renderer.php:33
Stack trace:
#0 /srv/htdocs/wp-content/plugins/mailpoet/vendor/mailpoet/email-editor/src/Engine/Renderer/class-renderer.php(32): MailPoet\EmailEditor\Engine\Renderer\ContentRenderer\Content_Renderer->render(Object(WP_Post), NULL)
#1 /srv/htdocs/wp-content/plugins/mailpoet/lib/Newsletter/Renderer/Renderer.php(95): MailPoet\EmailEditor\Engine\Renderer\Renderer->render(Object(WP_Post), 'Subject-new edi...', '', 'en-US', '<meta name="rob...')
#2 /srv/htdocs/wp-content/plugins/mailpoet/lib/Newsletter/Renderer/Renderer.php(85): MailPoet\Newsletter\Renderer\Renderer->_render(Object(MailPoet\Entities\NewsletterEntity), NULL, 'html', true, 'Subject-new edi...')
#3 /srv/htdocs/wp-content/plugins/mailpoet/lib/Newsletter/ViewInBrowser/ViewInBrowserRenderer.php(84): MailPoet\Newsletter\Renderer\Renderer->renderAsPreview(Object(MailPoet\Entities\NewsletterEntity), 'html')
#4 /srv/htdocs/wp-content/plugins/mailpoet/lib/Newsletter/ViewInBrowser/ViewInBrowserController.php(62): MailPoet\Newsletter\ViewInBrowser\ViewInBrowserRenderer->render(true, Object(MailPoet\Entities\NewsletterEntity), Object(MailPoet\Entities\SubscriberEntity), NULL)
#5 /srv/htdocs/wp-content/plugins/mailpoet/lib/Router/Endpoints/ViewInBrowser.php(32): MailPoet\Newsletter\ViewInBrowser\ViewInBrowserController->view(Array)
#6 [internal function]: MailPoet\Router\Endpoints\ViewInBrowser->view(Array)
#7 /srv/htdocs/wp-content/plugins/mailpoet/lib/Router/Router.php(74): call_user_func(Array, Array)
#8 /srv/htdocs/wp-content/plugins/mailpoet/lib/Config/Initializer.php(493): MailPoet\Router\Router->init()
#9 /srv/htdocs/__wp__/wp-includes/class-wp-hook.php(324): MailPoet\Config\Initializer->postInitialize('')
#10 /srv/htdocs/__wp__/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters(NULL, Array)
#11 /srv/htdocs/__wp__/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#12 /srv/htdocs/__wp__/wp-settings.php(722): do_action('wp_loaded')
#13 /srv/htdocs/wp-config.php(85): require_once('/srv/htdocs/__w...')
#14 /srv/htdocs/__wp__/wp-load.php(55): require_once('/srv/htdocs/wp-...')
#15 /srv/htdocs/__wp__/wp-blog-header.php(13): require_once('/srv/htdocs/__w...')
#16 /srv/htdocs/__wp__/index.php(17): require('/srv/htdocs/__w...')
#17 {main}
  thrown in /srv/htdocs/wp-content/plugins/mailpoet/vendor/mailpoet/email-editor/src/Engine/Renderer/ContentRenderer/class-content-renderer.php on line 33
```

costasovo committed:
Fix the Gutenberg version check to use the correct version

[MAILPOET-6367]

costasovo committed:
Refactor editor dependencies notice logic to extra class

Having the logic in a class will make it easier to reuse in multiple
places like the preview page.
[MAILPOET-6367]

costasovo committed:
Display dependency notice also for email previews

[MAILPOET-6367]

costasovo commented:
@triple0t Thank you for the review, and I'm very sorry I submitted a PR with so many issues. I was a bit in a rush in those days in between holidays, and I didn't realize that disabling initialization would break the notices.

I added three new commits and removed the commit where I previously completely skipped initialization when dependencies were not met. When the editor is not initialized, it doesn't register the post types, and it causes the issue with the unknown post type.

For the custom preview, I also added the redirect to newsletters, and I've slightly altered the text in the notice. In case an anonymous user somehow gets the preview URL, the notice will be displayed instead of the preview.

triple0t commented:
Awesome. Thank you for fixing the issue. @costasovo 

> I'm very sorry I submitted a PR with so many issues. I was a bit in a rush in those days in between holidays, and I didn't realize that disabling initialization would break the notices.

Not a problem. I totally understand.

> For the custom preview, I also added the redirect to newsletters, and I've slightly altered the text in the notice. In case an anonymous user somehow gets the preview URL, the notice will be displayed instead of the preview.

With regards to the preview, I still get an error when using WP 6.6.2 and Gutenberg 19.9.0. 
* I get an error when I click on the preview link on the Newsletter list page
* I get an error when I click on "Preview in new tab" within the editor

Using WP 6.6.2 and Gutenberg 19.9.0:
* I'm able to create a Newsletter and edit it
* BUT, I am not able to edit (or open) a Newsletter created with a different version of WP (WP 6.7) -- The editor loads but gets stuck on the loading icon (I don't think we need to worry about this option)

Logs
```
07-Jan-2025 14:01:34 UTC] PHP Fatal error:  Uncaught Error: Call to undefined method MailPoet\EmailEditor\Engine\PersonalizationTags\HTML_Tag_Processor::set_modifiable_text() in /srv/htdocs/wp-content/plugins/mailpoet/vendor/mailpoet/email-editor/src/Engine/class-personalizer.php:31
Stack trace:
#0 /srv/htdocs/wp-content/plugins/mailpoet/lib/Newsletter/ViewInBrowser/ViewInBrowserRenderer.php(110): MailPoet\EmailEditor\Engine\Personalizer->personalize_content('<!DOCTYPE html>...')
#1 /srv/htdocs/wp-content/plugins/mailpoet/lib/Newsletter/ViewInBrowser/ViewInBrowserController.php(71): MailPoet\Newsletter\ViewInBrowser\ViewInBrowserRenderer->render(true, Object(MailPoet\Entities\NewsletterEntity), Object(MailPoet\Entities\SubscriberEntity), NULL)
#2 /srv/htdocs/wp-content/plugins/mailpoet/lib/Router/Endpoints/ViewInBrowser.php(32): MailPoet\Newsletter\ViewInBrowser\ViewInBrowserController->view(Array)
#3 [internal function]: MailPoet\Router\Endpoints\ViewInBrowser->view(Array)
#4 /srv/htdocs/wp-content/plugins/mailpoet/lib/Router/Router.php(74): call_user_func(Array, Array)
#5 /srv/htdocs/wp-content/plugins/mailpoet/lib/Config/Initializer.php(493): MailPoet\Router\Router->init()
#6 /srv/htdocs/__wp__/wp-includes/class-wp-hook.php(324): MailPoet\Config\Initializer->postInitialize('')
#7 /srv/htdocs/__wp__/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters(NULL, Array)
#8 /srv/htdocs/__wp__/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#9 /srv/htdocs/__wp__/wp-settings.php(722): do_action('wp_loaded')
#10 /srv/htdocs/wp-config.php(85): require_once('/srv/htdocs/__w...')
#11 /srv/htdocs/__wp__/wp-load.php(55): require_once('/srv/htdocs/wp-...')
#12 /srv/htdocs/__wp__/wp-blog-header.php(13): require_once('/srv/htdocs/__w...')
#13 /srv/htdocs/__wp__/index.php(17): require('/srv/htdocs/__w...')
#14 {main}
  thrown in /srv/htdocs/wp-content/plugins/mailpoet/vendor/mailpoet/email-editor/src/Engine/class-personalizer.php on line 31
```

triple0t reviewed (commented):
Please check the comment here: https://github.com/mailpoet/mailpoet/pull/6018#issuecomment-2575406870

costasovo committed:
Update the dependency check logic and the message

We actually need WP 6.7 and above, no matter what Gutenberg version
is installed. This commit updates the check logic and the warning
message to match this requirement.
[MAILPOET-6367]

costasovo commented:
> Using WP 6.6.2 and Gutenberg 19.9.0:
> - I'm able to create a Newsletter and edit it
> - BUT, I am not able to edit (or open) a Newsletter created with a different version of WP (WP 6.7) -- The editor loads but gets stuck on the loading icon (I don't think we need to worry about this option)

@triple0t Thanks for checking that! I missed that for rendering; we need functionality from the WP core. So, no matter what Gutenberg version is installed, we require WP 6.7. I think it makes no sense to allow editing when we can't render it so I updated the logic of the check that WP needs to be always higher or equal to 6.7, and additionally, we check if there isn't an old Gutenberg version active. I've also updated the message.

triple0t commented on packages/php/email-editor/src/Engine/class-dependency-check.php:
```diff
@@ -0,0 +1,55 @@
+	 * Checks if the WordPress version is supported.
+	 */
+	private function is_gutenberg_version_compatible(): bool {
+		$plugin_data = get_plugin_data( WP_PLUGIN_DIR . '/gutenberg/gutenberg.php' );
```
Please use `get_plugin_data( WP_PLUGIN_DIR . '/gutenberg/gutenberg.php' , false, false)`

This ensures the data is not translated, which is not required in this context (and it's recommended), and it prevents the weird translation not loaded error in WP 6.7

triple0t reviewed (commented):
Thank you for making the change. 

If we need WP 6.7 regardless of the version of the Gutenberg plugin installed, why check for it?

We can simplify the check condition by removing the Gutenberg plugin version check. It currently serves no purpose.

costasovo committed:
Remove the Gutenberg version check from editor dependencies

MailPoet itself requires 6.6 to activate, so the oldest Gutenberg version we could get is 19.6, which is the first that allows WP 6.6 and could be active together with MailPoet. So, we are safe to remove the check for the Gutenberg version for now.
[MAILPOET-6367]

costasovo commented:
> We can simplify the check condition by removing the Gutenberg plugin version check. It currently serves no purpose.

@triple0t That's a great point! I kept it because, in theory, someone with a new WordPress could downgrade Gutenberg by using an older Gutenberg plugin, but it is really an edge case. On top of that, I can see that MailPoet itself requires WP 6.6 to activate so the oldest Gutenberg version we could get is 19.6, which is the first that allows WP 6.6. So, we are safe to remove it for now.

I removed the check in https://github.com/mailpoet/mailpoet/pull/6018/commits/5979ad217d83eebea7fea0a4fd786f14eb8f529a

triple0t reviewed (approved):
LGTM. Thanks for implementing the update
